MSAL/src-internal/ios/MSALLegacySharedAccount.m

```diff
@@ -0,0 +1,211 @@
+    }
+    
+    NSString *appBundleId = [[NSBundle mainBundle] bundleIdentifier];
+    
+    NSMutableDictionary *jsonDictionary = [NSMutableDictionary new];
+    jsonDictionary[@"id"] = [[NSUUID UUID] UUIDString];
+    jsonDictionary[@"environment"] = @"PROD";
+    
+    if (accountVersion == MSALLegacySharedAccountVersionV3)
+    {
+        jsonDictionary[@"originAppId"] = appBundleId;
+    }
+    
+    jsonDictionary[@"signInStatus"] = @{appBundleId : @"SignedIn"};
+    jsonDictionary[@"username"] = account.username;
+    jsonDictionary[@"additionalProperties"] = @{@"createdBy": appName};
+    [jsonDictionary addEntriesFromDictionary:[self claimsFromMSALAccount:account claims:claims]];
+    return [self initWithJSONDictionary:jsonDictionary error:error];
+}
+
+#pragma mark - Match
+
+- (BOOL)matchesParameters:(MSALAccountEnumerationParameters *)parameters
+{
+    if (parameters.needsAssociatedRefreshToken)
+    {
+        NSString *appIdentifier = [[NSBundle mainBundle] bundleIdentifier];
+        NSString *signinStatus = _signinStatusDictionary[appIdentifier];
+        
+        if (![signinStatus isEqualToString:@"SignedIn"])
+        {
+            return NO;
+        }
+        
+        return YES;
+    }
+    
+    return YES;
+}
+
+#pragma mark - Update
+
+- (BOOL)updateAccountWithMSALAccount:(id<MSALAccount>)account
+                     applicationName:(NSString *)appName
+                           operation:(MSALLegacySharedAccountWriteOperation)operation
+                      accountVersion:(MSALLegacySharedAccountVersion)accountVersion
+                               error:(NSError **)error
+{
+    if (accountVersion == MSALLegacySharedAccountVersionV1)
+    {
+        return YES;
+    }
+    
+    NSMutableDictionary *oldDictionary = [self.jsonDictionary mutableCopy];
+    NSString *appIdentifier = [[NSBundle mainBundle] bundleIdentifier];
+    
+    if (appIdentifier)
+    {
+        NSMutableDictionary *signinDictionary = [NSMutableDictionary new];
+        [signinDictionary addEntriesFromDictionary:_signinStatusDictionary];
+        
+        NSString *signinState = nil;
+        
+        switch (operation) {
+            case MSALLegacySharedAccountRemoveOperation:
+                signinState = @"SignedOut";
+                break;
+            case MSALLegacySharedAccountUpdateOperation:
+                signinState = @"SignedIn";
+                break;
+                
+            default:
+                return NO;
+        }
+        
+        signinDictionary[appIdentifier] = signinState;
+        oldDictionary[@"signInStatus"] = signinDictionary;
+    }
+    
+    NSDictionary *additionalAccountInfo = [self.jsonDictionary msidObjectForKey:@"additionalProperties" ofClass:[NSDictionary class]];
+    NSMutableDictionary *mutableAdditionalInfo = [NSMutableDictionary new];
+    [mutableAdditionalInfo addEntriesFromDictionary:additionalAccountInfo];
+    
+    mutableAdditionalInfo[@"updatedBy"] = appName;
```

Comment:
nit: Should we create constants for these keys "updatedBy" , "updatedAt" , e.t.c

Reply:
I'm personally not very found of constants when they're only used in one place. It's much easier to read when the actual string literal is used. If it's reused in multiple places, yes.